text/0006-equivocation-catastrophes.md

```diff
@@ -0,0 +1,111 @@
+
+As the last step, agree on the initial set of validators for the restarted network. The equivocators should be slashed, but there may be reasons to also exclude other validators, or even to start with a set of validators that has nothing in common with the validators from the stopped network. Once agreement has been reached, apply necessary changes to the pre-initial state: modify the validator stakes and change the entries in the auction contract. After this is applied, the resulting state will be the state used for the restart.
+
+After the users collectively decide upon the new initial state and the new initial set of validators, the network can be restarted.
+
+### Restarting the network
+
+The process of restarting the network would be very similar to the process of starting it. In its essence, it is just like starting the network from scratch, but with a few differences:
+
+- There is a different set of genesis validators - the new validators chosen by the community.
+- There is a different initial global state: it is not empty anymore, but it is the initial state prepared before.
+- Other details like the protocol version can differ as well, but they are not of the main concern for the restart process itself.
+
+The agreed upon set of validators and the initial state would be included in the chainspec. The "restart-chainspec" could be distributed among the validators the same way the initial chainspec for the network was. The files containing the new initial state could be distributed alongside the chainspec, or the nodes could download it from each other after they are started (which, of course, requires at least one node to possess the full copy of the state).
+
+## Reference-level explanation
+
+[reference-level-explanation]: #reference-level-explanation
+
+Here we will just summarize the exact steps to be performed when an equivocation catastrophe happens.
+
+1. An equivocation catastrophe is detected. Once a single validator detects it, we can expect all the validators to detect it eventually as well (even if just due to them receiving the evidence from the validator that detected it first).
+2. Consensus gets suspended, though validators continue to gossip. The users (validator nodes' owners) receive notice about the catastrophe and start the process of social consensus.
+3. The users agree about the new initial global state and the new initial set of validators. The new initial global state is manually constructed based on social consensus.
+4. A chainspec for the restart is manually constructed and agreed upon. This includes:
+    - the initial set of validators for the restarted network,
+    - the initial global state hash for the restarted network
```

Comment:
As mentioned above, I think this initial global state hash (including the replayed deploys and changed validator set) should not be given explicitly, but implicitly (just like the actual initial state hash when starting a _new_ network is also only given implicitly). Otherwise we need to provide special software to compute it.

Reply:
I think we will need special software to modify the initial global state, anyway (to replay chosen transactions and to modify the validator set). This software could just as well output the state hash after everything's done.

Reply:
I don't think it needs to be separate software, though. It could just be part of the node: Just like it currently computes the "initial" global state using the contracts configured in the chainspec, it should compute the new global state after such a restart, based on a _block_ hash (the last finalized block), a list of additional block and/or deploy hashes, and possibly some upgrade-like code.